docs/source/configuration.rst
Cerise Configuration
====================

Introduction
------------

Cerise takes configuration information from various sources, with some
overriding others. This page describes the configuration files and what can be
configured in them.

.. _main_configuration:

Main configuration file
-----------------------

The main configuration file is located at ``conf/config.yml``, and contains
general configuration information for the Cerise service in YAML format. It
looks as follows::

  database:
    file: run/cerise.db

  logging:
    file: /var/log/cerise/cerise_backend.log
    level: INFO

  pidfile: run/cerise_backend.pid

  client-file-exchange:
    store-location-service: file:///tmp/cerise_files
    store-location-client: file:///tmp/cerise_files

  rest-service:
    base-url: http://localhost:29593
    hostname: 127.0.0.1
    port: 29593

Cerise uses SQLite to persistently store the jobs that have been submitted to
it. SQLite databases consist of a single file, the location of which is given by
the ``file`` key under ``database``.

Logging output is configured under the ``logging`` key. Make sure that the user
that Cerise runs under has write access to the given path. If you want to log to
/var/log without giving Cerise root rights, making the specified log file on
beforehand and then giving ownership to the user Cerise runs under works well.
Or you can make a subdirectory and give the user access to that.

The ``pidfile`` key specifies a path to a file into which Cerise's process
identifier (PID) is written. This can be used to shut down a running service,
i.e. ``kill <pid>`` will cleanly shut down Cerise.

Under ``client-file-exchange``, the means of communicating files between Cerise
and its users is configured. Communication is done using a shared folder
accessible to both the users and the Cerise service. If Cerise is running
locally, both parties have access to the same file system, and see the shared
folder in the same location. Thus, ``store-location-service`` and
``store-location-client`` both point to the same on-disk directory.

If the Cerise service does not share a file system with the client, then a
directory on the Cerise server must be made available to the client, e.g. via
WebDAV. In this case, client and service access the same directory using
different URLs, e.g.

::

  client-file-exchange:
    store-location-service: file:///home/webdav/files
    store-location-client: http://localhost:29593/files

The user is expected to submit references to files that start with the URL in
``store-location-client``, Cerise will then fetch the corresponding files from
the directory specified in ``store-location-service``.

``store-location-client`` can be overridden by specifying the environment
variable CERISE_STORE_LOCATION_CLIENT. If you want to run multiple Cerise
instances in containers, simultaneously, then you need to remap the ports on
which they are available to avoid collisions. With this environment variable,
the port can be easily injected into the container, removing the need to have
a different image for each container. Cerise Client uses this functionality.

Finally, key ``rest-service`` has the hostname and port on which the REST
service should listen, as well as the external URL on which it is available.
If you want the service to be available to the outside
world, this should be the IP address of the network adaptor to listen on, or
``0.0.0.0`` to listen on all adaptors. Note that a service running inside a
Docker container needs to have ``0.0.0.0`` for it to be accessible from outside
the container.

Since the service needs to pass URLs to the client sometimes, it needs to know
at which URL it is available to the client. This is specified by ``base-url``,
which should contain the first part of the URL to the REST API, before the
``/jobs`` part. Alternatively, you can set the CERISE_BASE_URL environment
variable to this value.

.. _compute-resource-configuration:

Compute resource configuration
------------------------------

Information on which compute resource to connect to, and how to transfer files
and submit jobs to it, is stored separately from the main service configuration,
to make it easier to create specialisations. Furthermore, to enable different
users to use the same specialised Cerise installation (e.g. Docker image),
credentials can be specified using environment variables. (Cerise Client uses
the latter method.) If you are making a specialisation that is to be shared with
others, do not put your credentials in this file!

Note: this file is somewhat outdated, but well be updated prior to the 1.0 release.

API configuration file
......................

The API configuration file is located in ``api/config.yml``, and has the following
format::

  compute-resource:
    credentials:
      username: None
      password: None
      certfile: None
      passphrase: None

    files:
      credentials:
        username: None
        password: None
        certfile: None
        passphrase: None

      protocol: local
      location: None
      path: /home/$CERISE_USERNAME/.cerise

    jobs:
      credentials:
        username: None
        password: None
        certfile: None
        passphrase: None

      protocol: local
      location: None
      scheduler: none

      queue-name: None      # cluster default
      slots-per-node: None  # cluster default
      cores-per-node: 32
      scheduler-options: None
      cwl-runner: $CERISE_API_FILES/cerise/cwltiny.py

    refresh: 10

This file describes the compute resource and how to connect to it. Under the
``files`` key, file access (staging) is configured, while the ``jobs`` key has
settings on how to submit jobs. ``credentials``, and keys ``username``,
``password``, ``certfile`` and ``passphrase`` occurring throughout, refer to
credentials, and will be discussed below. Keys may be omitted if they are not
needed, e.g. ``location`` may be omitted if ``protocol`` is ``local``, in which
case credentials may also me left out.

For file staging, a protocol, location and path may be specified.  Supported
protocols are ``file``, ``sftp``, ``ftp``, or ``webdav``, where ``file`` refers
to direct access to the local file system.

``location`` provides the host name to connect to; to run locally, this may be
omitted or empty. ``path`` configures the remote directory where Cerise will put
its files. It may contain the string ``$CERISE_USERNAME``, which will be
replaced with the user account name that the service is using. This is useful if
you want to put Cerise's files into the users home directory, e.g.
``/home/$CERISE_USERNAME/.cerise`` (which is the default value). Note that
user's home directories are not always in ``/home`` on compute clusters, so be
sure to check this.

Job management is configured under the ``jobs`` key. Here too a protocol may be
given, as well as a location, and a few other settings can be made.

For job management, the protocol can be ``local`` (default) or ``ssh``. If the
``local`` protocol is selected, ``location`` is ignored, and jobs are run
locally. For the ``ssh`` protocol, ``location`` is the name of the host,
optionally followed by a colon and a port number (e.g. ``example.com:2222``).

Jobs can be run directly or via a scheduler. To run jobs directly, either on the
local machine or on some remote host via SSH, set the scheduler to ``none``.
Other valid values for ``scheduler`` are ``slurm``, ``torque`` and
``gridengine`` to submit jobs to the respective job management system.

If jobs need to be sent to a particular queue, then you can pass the queue name
using the corresponding option; if it is not specified, the default queue is
used. If one or more of your steps start MPI jobs, then you may want to set the
number of MPI slots per node via ``slots-per-node`` for better performance. If
you need to specify additional scheduler options to e.g. select a GPU node, you
can do so using e.g. ``scheduler-options: "-C TitanX --gres=gpu:1"``. Ideally,
it would be possible to specify this in the CWL file for the step, but support
for this in CWL is partial and in-development, and Cerise does not currently
support this. Users can specify the number of cores to run on using a CWL
ResourceRequirement, but Cerise always allocates whole nodes. It therefore needs
to know the number of cores in each node, which you should specify using
``cores-per-node``.

Finally, ``cwl-runner`` specifies the remote path to the CWL runner. It defaults
to ``$CERISE_API_FILES/cerise/cwltiny.py``, which is Cerise's included simple
CWL runner. ``$CERISE_API_FILES`` will be substituted for the appropriate remote
directory by Cerise. See :doc:`Specialising Cerise <specialising>` for more
information.

Cerise will regularly poll the compute resource it is connected to, to check if
any of the running jobs have finished. The ``refresh`` setting can be used to
set the minimum interval in seconds between checks, so as to avoid putting too
much load on the machine.

Credentials may be put into the configuration file as indicated. Valid
combinations are:

- No credentials at all (for running locally)
- Only a username
- A username and a password
- A username and a certificate file
- A username, a certificate file, and a passphrase

If the credentials to use for file access and job management are the same, then
you should list them under ``credentials`` and omit them in the other locations.
If different credentials are needed for files and jobs, then a ``credentials``
block can be specified under ``files`` and ``jobs`` respectively. Credentials
listed here may be overridden by environment variables, as described below.


Environment variables
.....................

Cerise checks a set of environment variables for credentials. If found, they
override the settings in the configuration file. These variables are:

General credentials

- CERISE_USERNAME
- CERISE_PASSWORD
- CERISE_CERTFILE
- CERISE_PASSPHRASE

Credentials for file access

- CERISE_FILES_USERNAME
- CERISE_FILES_PASSWORD
- CERISE_FILES_CERTFILE
- CERISE_FILES_PASSPHRASE

Credentials for job management

- CERISE_JOBS_USERNAME
- CERISE_JOBS_PASSWORD
- CERISE_JOBS_CERTFILE
- CERISE_JOBS_PASSPHRASE

As in the configuration file, specific credentials go before general ones.
Cerise will first try a specific environment variable (e.g.
CERISE_JOBS_USERNAME), then the corresponding specific configuration file entry
(under ``jobs``), then a generic environment variable (e.g. CERISE_USERNAME),
and finally the generic configuration file entry (under ``credentials``).

It does this for each of the four credential components separately, then uses
the first complete combination from the top down to connect:

- username + certfile + passphrase
- username + certfile
- username + password
- username
- <no credentials>



CHANGELOG.rst
###########
Change Log
###########

All notable changes to this project will be documented in this file.
This project adheres to `Semantic Versioning <http://semver.org/>`_.

0.2.0
*****

Added
-----

* Better HPC support
  * Extra scheduler options (e.g. to select a GPU node)
  * Allow specifying the number of cores, per step
  * Allow specifying a time limit, per step and/or per job

* Improved specialisations
  * Allow multiple projects per specialisation
  * Versioning of specialisations
  * Automatic remote installation of updates
  * Rewritten documentation

Fixed
-----

* Usability
  * Improved job logs
  * Improved server log
  * Incremental remote job log updates
  * Log level configurable via environment variable

* Robustness
  * Automatically reconnect if the network connection drops
  * Better error-handling and retrying throughout
  * Better error messages
  * Database recovery in the event of a crash

* Technical improvements
  * Faster data transfer
  * Cleaner code, using Cerulean instead of Xenon 1
  * Better code documentation
  * Improved and cleaned up unit tests
  * Rebuilt integration test
  * Type annotations and checking using mypy

0.1.0
*****

Initial release


cerise/config.py
import logging
import os
import traceback
import urllib
from typing import Any, Dict, Optional, cast

import cerulean
import yaml

_remote_file_system = None


class Config:
    def __init__(self, config: Dict[str, Any],
                 api_config: Dict[str, Any]) -> None:
        """Create a configuration object.

        Args:
            config (dict): A main configuration dict.
            api_config (dict): An API configuration dict.
        """
        self._logger = logging.getLogger(__name__)
        """Logger: The logger for this class."""
        self._config = config
        """The main configuration dictionary."""
        self._api_config = api_config
        """The API configuration dictionary."""
        self._cr_config = dict()  # type: Dict[str, Any]
        """The compute-resource part of the main config."""

        if 'compute-resource' in self._api_config:
            self._cr_config = cast(dict, self._api_config['compute-resource'])

    def _get_credential_variable(self, kind: str, name: str) -> Optional[str]:
        def have_config(kind: str, name: str) -> bool:
            if kind == '':
                return ('credentials' in self._cr_config
                        and name in self._cr_config['credentials'])
            return (kind in self._cr_config
                    and 'credentials' in self._cr_config[kind]
                    and name in self._cr_config[kind]['credentials'])

        def get_config(kind: str, name: str) -> str:
            if kind == '':
                return self._cr_config['credentials'][name]
            return self._cr_config[kind]['credentials'][name]

        def env_var(kind: str, name: str) -> str:
            if kind == '':
                return '_'.join(['CERISE', name.upper()])
            else:
                return '_'.join(['CERISE', kind.upper(), name.upper()])

        def have_env(kind: str, name: str) -> bool:
            return env_var(kind, name) in os.environ

        def get_env(kind: str, name: str) -> str:
            return os.environ[env_var(kind, name)]

        value = None
        if have_config('', name):
            value = get_config('', name)
        if have_env('', name):
            value = get_env('', name)
        if have_config(kind, name):
            value = get_config(kind, name)
        if have_env(kind, name):
            value = get_env(kind, name)
        return value

    def _get_credential(self, kind: str) -> Optional[cerulean.Credential]:
        """
        Create a Cerulean Credential given the configuration, and return
        it together with the username.

        Args:
            kind (str): Either 'files' or 'jobs'.

        Returns:
            (cerulean.Credential): The credential to use for connecting
        """
        username = self._get_credential_variable(kind, 'username')
        password = self._get_credential_variable(kind, 'password')
        certfile = self._get_credential_variable(kind, 'certfile')
        passphrase = self._get_credential_variable(kind, 'passphrase')

        credential = None  # type: Optional[cerulean.Credential]
        if username and certfile:
            credential = cerulean.PubKeyCredential(username, certfile,
                                                   passphrase)
        elif username and password:
            credential = cerulean.PasswordCredential(username, password)

        return credential

    def close_file_systems(self) -> None:
        """Close any open connections and free resources.

        This function is to be called on shutdown, to ensure that the
        remote file system managed by Config is shut down properly.
        """
        global _remote_file_system
        if _remote_file_system is not None:
            _remote_file_system.close()

    def get_service_host(self) -> str:
        """
        Return the host interface Cerise should listen on.

        Returns:
            str: The IP address of the interface to listen on.
        """
        rs_config = self._config.get('rest-service', {})
        return rs_config.get('hostname', '127.0.0.1')

    def get_service_port(self) -> int:
        """
        Return the port on which Cerise should listen.

        Returns:
            int: The port number to listen on.
        """
        rs_config = self._config.get('rest-service', {})
        return int(rs_config.get('port', 29593))

    def get_username(self, kind: str) -> Optional[str]:
        """
        Return the username used to connect to the specified kind of resource.

        Args:
            kind (str): Either 'files' or 'jobs'

        Returns:
            (str): The configured username
        """
        return self._get_credential_variable(kind, 'username')

    def get_scheduler(self,
                      run_on_head_node: bool = False) -> cerulean.Scheduler:
        """
        Returns a scheduler as configured by the user.

        Args:
            run_on_head_node (bool): If True, will create a scheduler using \
            the ssh adaptor instead of the configured one if the configured \
            adaptor is a cluster scheduler (i.e. slurm, torque or gridengine).

        Returns:
            (cerulean.Scheduler): A new scheduler
        """
        if 'jobs' not in self._cr_config:
            protocol = 'local'
            location = None
            scheduler_type = 'directgnu'
        else:
            protocol = self._cr_config['jobs'].get('protocol', 'local')
            location = self._cr_config['jobs'].get('location')
            scheduler_type = self._cr_config['jobs'].get(
                'scheduler', 'directgnu')

        if run_on_head_node:
            scheduler_type = 'directgnu'

        credential = self._get_credential('jobs')
        terminal = cerulean.make_terminal(protocol, location, credential)
        scheduler = cerulean.make_scheduler(scheduler_type, terminal)
        return scheduler

    def get_file_system(self) -> cerulean.FileSystem:
        """
        Returns a remote file system as configured by the user.

        Returns:
            (cerulean.FileSystem) A new filesystem
        """
        global _remote_file_system
        if _remote_file_system is None:
            if 'files' not in self._cr_config:
                protocol = 'local'
                location = None
            else:
                protocol = self._cr_config['files'].get('protocol', 'local')
                location = self._cr_config['files'].get('location')

            credential = self._get_credential('files')
            self._logger.debug(
                ('protocol: {}, location: {}, credential: {}').format(
                    protocol, location, credential))

            _remote_file_system = cerulean.make_file_system(
                protocol, location, credential)

        return _remote_file_system

    def get_remote_cwl_runner(self) -> str:
        """
        Returns the configured remote path to the CWL runner to use.

        No macro substitution is done; this gives the configured path as-is.

        Returns:
            (str): The path.
        """
        default = '$CERISE_API/cerise/files/cwltiny.py'
        if 'jobs' not in self._cr_config:
            return default
        return self._cr_config['jobs'].get('cwl-runner', default)

    def get_basedir(self) -> cerulean.Path:
        """
        Returns the configured remote base directory to use.

        Returns:
            (str): The remote path to the base directory.
        """
        basedir = '/home/$CERISE_USERNAME/.cerise'
        if 'files' in self._cr_config:
            basedir = self._cr_config['files'].get('path', basedir)

        username = self.get_username('files')
        if username is not None:
            basedir = basedir.replace('$CERISE_USERNAME', username)
        basedir = basedir.strip('/')
        return self.get_file_system() / basedir

    def get_queue_name(self) -> Optional[str]:
        """
        Returns the name of the queue to submit jobs to, or None if no
        queue name was configured.

        Returns:
            (Union[str,None]): The queue name.
        """
        if 'jobs' not in self._cr_config:
            return None
        return self._cr_config['jobs'].get('queue-name')

    def get_slots_per_node(self) -> int:
        """
        Returns the configured number of MPI slots per node.

        Returns:
            (int): The number of slots to use.
        """
        default = 1
        if 'jobs' not in self._cr_config:
            return default
        return self._cr_config['jobs'].get('slots-per-node', default)

    def get_scheduler_options(self) -> Optional[str]:
        """Returns the additional scheduler options to use.

        Returns:
            (str): The options as a single string.
        """
        if 'jobs' not in self._cr_config:
            return None
        return self._cr_config['jobs'].get('scheduler-options', None)

    def get_cores_per_node(self) -> int:
        """Returns the number of cores per node.

        This depends on the available compute hardware, and should be
        configured in the specialisation. The incoming workflow
        specifies a number of cores, but we reserve nodes, so we need
        to convert.

        The default is 32, which is probably more than what you have,
        as a result of which we'll allocate fewer nodes than the user
        specified if no value is given. That'll slow things down, but
        at least we won't be burning core hours needlessly.

        Returns:
            (int): The number of cores per node on this machine.
        """
        if 'jobs' not in self._cr_config:
            return 32
        return self._cr_config['jobs'].get('cores-per-node', 32)

    def get_remote_refresh(self) -> float:
        """
        Returns the interval in between checks of the remote job \
        status, in seconds.

        Returns:
            (float): How often to check remote job status.
        """
        return self._cr_config.get('refresh', 60.0)

    def get_database_location(self) -> str:
        """
        Returns the local path to the database file.

        Returns:
            (str): The path.

        Raises:
            KeyError: No database path was set.
        """
        return self._config['database']['file']

    def get_pid_file(self) -> Optional[str]:
        """
        Returns the location of the PID file, if any.

        Returns:
            (Union[str,None]): The configured path, or None
        """
        return self._config.get('pidfile')

    def has_logging(self) -> bool:
        """
        Returns if logging is configured.

        Returns:
            True iff a logging section is available in the configuration.
        """
        return 'logging' in self._config

    def get_log_file(self) -> str:
        """
        Returns the configured path for the log file. Use has_logging()
        to see if logging has been configured first.

        Returns:
            (str): The path.
        """
        return self._config['logging'].get(
            'file', '/var/log/cerise/cerise_backend.log')

    def get_log_level(self) -> int:
        """
        Returns the configured log level. Use has_logging() to see if
        logging has been configured first.

        Returns:
            (int): The log level, following Python's built-in logging \
                    library.
        """
        loglevel_str = 'INFO'
        if 'logging' in self._config:
            loglevel_str = self._config['logging'].get('level', 'INFO')
        if 'CERISE_LOG_LEVEL' in os.environ:
            loglevel_str = os.environ['CERISE_LOG_LEVEL'].strip()
        loglevel = getattr(logging, loglevel_str.upper(), None)
        return loglevel

    def get_base_url(self) -> str:
        """Returns the service's base url.

        This is the URL of the REST API, before the /jobs part, e.g. if
        listing jobs is done by a GET to http://localhost/jobs, then
        this should be set to http://localhost. Obtained from the
        configuration file or the CERISE_BASE_URL environment variable.
        """
        if 'CERISE_BASE_URL' in os.environ:
            return os.environ['CERISE_BASE_URL']
        rs_config = self._config.get('rest-service', {})
        return rs_config.get('base-url', '')

    def get_store_location_service(self) -> cerulean.Path:
        """
        Returns the file exchange location access point for the service.

        Returns:
            The local base directory for file exchange with the client.

        Raises:
            KeyError: The location was not set.
        """
        url = self._config['client-file-exchange']['store-location-service']
        urlparts = urllib.parse.urlparse(url)
        if urlparts.scheme == 'file':
            return cerulean.LocalFileSystem() / urlparts.path
        elif urlparts.scheme == 'http':
            return cerulean.WebdavFileSystem(url) / ''
        else:
            raise RuntimeError('Config store-location-service contains an'
                               ' invalid scheme. Use file:// or http://.')

    def get_store_location_client(self) -> str:
        """
        Returns the file exchange location access point for the client.

        Returns:
            (str): A URL.

        Raises:
            KeyError: The location was not set.
        """
        if 'CERISE_STORE_LOCATION_CLIENT' in os.environ:
            return os.environ['CERISE_STORE_LOCATION_CLIENT']
        return self._config['client-file-exchange'].get(
            'store-location-client')


def make_config() -> Config:
    """Make a configuration object.

    Uses the configuration files and environment variables to determine
    the configuration.

    Returns:
        Config: The Cerise configuration.
    """
    config_file_path = 'conf/config.yml'
    try:
        with open(config_file_path) as config_file:
            config = yaml.safe_load(config_file)
    except Exception:
        print("Could not load main configuration, aborting.")
        print("Does the file exist, and is it valid YAML?")
        print(traceback.format_exc())
        quit(1)

    api_config_file_path = 'api/config.yml'
    try:
        with open(api_config_file_path) as api_config_file:
            api_config = yaml.safe_load(api_config_file)
    except Exception:
        print("Could not load API configuration, aborting.")
        print("Does the file exist, and is it valid YAML?")
        print(traceback.format_exc())
        quit(1)

    return Config(config, api_config)
